common/nmea_parser.h

```diff
@@ -0,0 +1,205 @@
+10) Units of antenna altitude, meters
+11) Geoidal separation, the difference between the WGS-84 earth
+ellipsoid and mean-sea-level (geoid), '-' means mean-sea-level below ellipsoid
+12) Units of geoidal separation, meters
+13) Age of differential GPS data, time in seconds since last SC104
+type 1 or 9 update, null field when DGPS is not used
+14) Differential reference station ID, 0000-1023
+15) Checksum
+*/
+
+#pragma once
+// standard includes
+#include<string>
+#include<vector>
+#include <sstream>
+
+// BoB includes
+#include "../third_party/units.h"
+#include "map_coordinate.h"
+
+namespace BoBRobotics
+{
+namespace GPS 
+{
+
+class GPSError
+  : public std::runtime_error
+{
+public:  
+    GPSError(const std::string &msg) : std::runtime_error(msg) { }
+};
+
+enum class GPSQuality
+{
+    INVALID = 0,  
+    GPSFIX  = 1,  
+    DGPSFIX = 2,
+    PPSFIX  = 3,
+    RTK     = 4,
+    FRTK    = 5
+};
+
+struct TimeStamp
+{
+    units::time::hour_t hour;
+    units::time::minute_t minute;
+    units::time::second_t second;
+    units::time::millisecond_t millisecond;
+    TimeStamp(){}
+    TimeStamp (const int &h, const int &m, const int &s, const int &ms)
+    : hour(h), minute(m), second(s), millisecond(ms) {}
+};
+
+struct GPSData {
+    BoBRobotics::MapCoordinate::GPSCoordinate coordinate;  // Latitude and longitude coordinate
+    units::length::meter_t altitude;                       // Altitude above ellipsoid
+    units::velocity::meters_per_second_t velocity;         // velocity
+    int numberOfSatelites;                                 // Currently observed number of satelites
+    double horizontalDilution;                             // horizontal dilution - lower value is better
+    GPSQuality gpsQuality;                                 // GPS quality indicator
+    TimeStamp time;                                        // time of measurement
+};
+
+class NMEAParser {
+
+    using second_t = units::time::second_t;
+    using minute_t = units::time::minute_t;
+    using hour_t = units::time::hour_t;
+    using millisecond_t = units::time::millisecond_t;
+    using meter_t = units::length::meter_t;
+    using degree_t = units::angle::degree_t;
+    using arcminute_t = units::angle::arcminute_t;
+    using meters_per_second_t = units::velocity::meters_per_second_t;
+
+    private:
+
+    static std::vector<std::string> parseNMEAstring(const std::string &textToParse, const std::string &NMEA_sentence_id) {
+
+        using namespace std;
+        const char delimiter = '$';      // sentences separated by [$]
+        const char w_delimiter = ',';    // elements separated by  [,]
+
+        vector<string> sentences,words;
+        istringstream split(textToParse);
+        // separating the text to sentences
+        
+        if (textToParse.empty()) throw GPSError("Empty string to parse");
+
+        for (string each; getline(split, each, delimiter); sentences.push_back(each));    
```

Comment:
Why do you need to read all the sentences when you only use one? Do the outer loop and check if ``each`` contains ``NMEA_sentence_id``